tests/data_classes.py

```diff
@@ -0,0 +1,642 @@
+    def recursive_snapshots(self):
+        immediate_snaps = [
+            s['id'].split('@')[1] for s in self.zfs.get_snapshots(self.path)
+        ]
+
+        return [
+            Snapshot(s['id'], self, self.zfs)
+            for s in self.zfs.get_snapshots_recursively(self.path)
+            if s['id'].split('@')[1] in immediate_snaps
+            # The last if is for logic copied from iocage
+        ]
+
+    @property
+    def exists(self):
+        return self.zfs.get_dataset_safely(self.path) is not None
+
+    @property
+    def is_template(self):
+        return 'iocage/templates/' in self.path
+
+    @property
+    def jid(self):
+        try:
+            return int(
+                subprocess.check_output(
+                    ['jls', '-j', f'ioc-{self.name.replace(".", "_")}'],
+                    stderr=subprocess.PIPE
+                ).decode('utf-8').split()[5]
+            )
+        except subprocess.CalledProcessError:
+            return None
+
+    @property
+    def running(self):
+        return self.jid is not None
+
+    @property
+    def config(self):
+        # TODO: Let's add validation for props as well in future
+        config = None
+        if os.path.exists(os.path.join(
+                self.absolute_path, self.DEFAULT_JSON_FILE
+        )):
+            with open(
+                os.path.join(self.absolute_path, self.DEFAULT_JSON_FILE), 'r'
+            ) as f:
+                try:
+                    config = json.loads(f.read())
+                except json.JSONDecodeError:
+                    pass
+
+        assert config is not None, f'Failed to read config.json for {self.name}'
+
+        return config
+
+    @property
+    def fstab(self):
+        # Let's return a list of fstab lines
+        fstab = None
+        if os.path.exists(os.path.join(
+            self.absolute_path, 'fstab'
+        )):
+            with open(os.path.join(self.absolute_path, 'fstab'), 'r') as f:
+                fstab = [
+                    l.split('#')[0].strip().replace('\t', ' ')
+                    for l in f.readlines() if l
+                ]
+
+        assert fstab is not None
+
+        return fstab
+
+    @property
+    def release(self):
+        return self.config.get('release', 'EMPTY')
+
+    @property
+    def is_thickconfig(self):
+        return 'CONFIG_TYPE' in self.config
+
+    @property
+    def is_basejail(self):
+        return self.config.get('basejail', 'no') == 'yes'
+
+    @property
+    def is_empty(self):
+        return self.config.get('release', 'EMPTY') == 'EMPTY'
+
+    @property
+    def is_thickjail(self):
+        return not self.root_dataset[
+            'properties'
+        ].get('origin', {}).get('value')
+
+    @property
+    def is_rcjail(self):
+        return self.config.get('boot', 'off') == 'on'
+
+    @property
+    def ip(self):
+        assert self.running is True
+        try:
+            # TODO: we should probably load interface from config and
```

Comment:
Since users can supply a custom vnet iface, we should copy iocage's logic here. As they might not be called epair0b, or if they have multiple interfaces.